### ISSUE TITLE
CFP: CiliumNetworkPolicy `toFQDNs.matchPattern` should allow wildcarding of the `.` character to allow multiple levels of subdomains

### ISSUE THREAD

**Issue by illrill** [cilium/cilium#22081]
## CiliumNetworkPolicy `toFQDNs.matchPattern` should allow wildcarding of the `.` character to allow multiple levels of subdomains

_2022-11-16: Updated issue description after discussions in comments. See the collapsed section in the end for original post_

**Is your feature request related to a problem?**
The current behavior of wildcards (`*`) in `toFQDNs.matchPattern` has the limitation that it does not include dots. In order to allow e.g. `example.com` and all of its subdomains (in multiple "levels"), e.g. `australia-east.example.com`, `api.australia-east.example.com`, `auth.api.australia-east.example.com`, I need to add four separate rules ([docs ref](https://docs.cilium.io/en/v1.12/policy/language/#dns-based));. 

```
  - toFQDNs:
    - matchPattern: "example.com"
    - matchPattern: "*.example.com"
    - matchPattern: "*.*.example.com"
    - matchPattern: "*.*.*.example.com"
```

**Describe the feature you'd like**
It should be possible to wildcard multiple level of subdomains for a given domain, i.e. the `.` character should be possible to wildcard. 

**Describe your proposed solution**
Introduce the option to use double asterixes `**` as a more "aggressive" wildcard in addition to the existing single asterix `*`.

_Current behavior_
- ``*`` within a domain allows 0 or more valid DNS characters, **except for the ``.`` separator**.
  * ``*.cilium.io`` will match ``sub.cilium.io`` but not ``sub.sub.cilium.io`` or  ``cilium.io``.
  *  ``part*ial.com`` will match ``partial.com`` and ``part-extra-ial.com``.
  * ``*cilium.io`` will match ``anothercilium.io`` but not ``sub.anothercilium.io`` or ``sub.cilium.com``.

_Suggested additional behavior_
- `**` within a domain allows 0 or more valid DNS characters, **including the `.` separator**. (New)
  * ``**.cilium.io`` will match ``sub.cilium.io`` and ``sub.sub.cilium.io``, but not ``cilium.io``.
  * ``part**ial.com`` will match ``partial.com`` and ``part-extra-ial.com``.
  * ``**cilium.io`` will match ``anothercilium.io`` and ``sub.anothercilium.io`` and ``sub.cilium.com``.

This way, to address the problem scenario, I could reduce my toFQDNs items into one `matchName` (or `matchPattern`) to allow the exact domain, and another `matchPattern` to allow all levels of subdomains to that domain:

```
  - toFQDNs:
    - matchName: "example.com"
    - matchPattern: "**.example.com"
```

<details>
  <summary>Original issue description</summary><br>

## Using `*` in CiliumNetworkPolicy `toFQDNs.matchPattern` should include all levels of subdomains

**Is your feature request related to a problem?**
The current behavior of wildcards (`*`) in `toFQDNs.matchPattern` has the limitation that it does not include dots. In order to allow e.g. `example.com` and all of its subdomains (in multiple "levels"), e.g. `australia-east.example.com`, `api.australia-east.example.com`, `auth.api.australia-east.example.com`, I need to add four separate rules:

```
  - toFQDNs:
    - matchPattern: "example.com"
    - matchPattern: "*.example.com"
    - matchPattern: "*.*.example.com"
    - matchPattern: "*.*.*.example.com"
```

From the documentation (https://docs.cilium.io/en/v1.12/policy/language/#dns-based):
> `*` within a domain allows 0 or more valid DNS characters, except for the `.` separator. `*.cilium.io` will match `sub.cilium.io` but not `cilium.io`.

**Describe the feature you'd like**

`*` in `toFQDNs.matchPattern` should match _zero to unlimited characters_ like regex does with `(.*)`. So that I only need two rules; one that matches the domain, and one that matches any level of subdomain to that domain:

```
  - toFQDNs:
    - matchPattern: "example.com"
    - matchPattern: "*.example.com"
```
</details>

**Comment by joestringer** [comment]
A couple of notes:
- During an early version of the design, this was possible, but then the original author decided it may be safer to avoid this. I vaguely recall some concern that wide-scoped rules could encourage administrators to write unsafe rules, for example by matching all subdomains it could make subdomain hijacking easier. At a glance I couldn't find the original rationale, but you may be able to find it by digging through old PRs related to the `toFQDNs` policies.
- Many users have written their rules against the current semantics, so there is no way that we could change the current `matchPattern` statements to reinterpret the `.` or `*` characters. If this is supported, it would need to be a new field.

**Comment by illrill** [comment]
Thanks for the notes Joe.

- Indeed it is probably safer considering a subdomain takeover scenario, but for administrators who prefer configuration simplicity over maximal safety, it can be considered somewhat limiting. It could be argued that this decision between safety and simplicity should be in the hands of the administrator and not limited by the product itself. Looking at how other types of "firewalls", e.g. [Azure Firewall](https://learn.microsoft.com/en-us/azure/firewall/firewall-faq#how-do-wildcards-work-in-target-urls-and-target-fqdns-in-application-rules) have chosen to do it, `*` matches _any character_, allowing for wildcarding multiple levels of subdomains.

- Good point regarding the breaking change. Perhaps rather than introducing a completely new field, a less intrusive option could be to introduce another allowed "type" of wildcard character. For example **double asterixes**:
  - `**` matches any character (including the `.`). Allows wildcarding of multiple levels of subdomains.
  - `*` matches any character except `.` (the way it is now). Allows wildcarding of a single subdomain.

**Comment by joestringer** [comment]
Broadly your arguments make sense to me, this seems like a quality-of-life improvement suggestion for users who are already accepting of the risk of wildcarded FQDN rules. I like the `**` suggestion since it is unlikely anyone is specifying this in their rules today, and yet it would provide a more explicit wildcarding option to cover this slightly different use case vs. the current semantics. It's also nice that it shouldn't require a new field like I had previously assumed.

**Comment by illrill** [comment]
Thank you for the feedback @joestringer. I updated the issue to reflect our discussions.

**Comment by illrill** [comment]
Can this be reopened @aanm @joestringer? It's a good feature 😃

**Comment by jrajahalme** [comment]
IMO `**` for matching multiple subdomains feels useful, but allowing it to match partial subdomain names seems very dangerous, especially if applied on the top-level domain or the next level down, as domain ownership seems murky. On the other hand, we already allow for `*.com` for example, so `**pple.com` is no less safe than that. But the fact that this matches `www.bad-apple.com` seems troubling.

Would limiting this to just multiple levels of subdomains, as indicated in the CFP name be acceptable?

**Comment by joestringer** [comment]
I think it's a reasonable limitation to require that `**` follows with `.` to indicate that it is only matching subdomains and not performing partial domain matches.

**Comment by phj-incom** [comment]
We have been testing the fix described at https://github.com/cilium/cilium/issues/22081#issuecomment-1767811688 for a week now without any issues. Our only concern is an increased memory usage for the CNI due to more IPs being held in memory due to the larger lookups.

Any chance this could get into a proper release?

**Comment by erikschul** [comment]
In reference to #35819, I will add the following

>I am also missing a top-level operator, that allows hostnames only, e.g. server1, but doesn't allow server1.com. Using a matchPattern * would have been correct, except that this was programmed as a special case that works as ** would.
Changing this would unfortunately break existing policies. Perhaps a different operator needs to be introduced, #, but it's obviously very ugly and unfortunate.

For this reason, I agree with @joestringer  that perhaps it is better to use a new field, and instead of the special case use of `matchPattern: *` use `matchPattern2: **` for the same behavior (wildcard including `.`), such that `*` is strictly a wildcard excluding `.` for consistency, i.e. `matchPattern2: *` would match `example` but not `example.com`.

**Comment by joestringer** [comment]
I don't think a new field should be necessary if the operator is different, such as `**`.

**Comment by erikschul** [comment]
@joestringer The problem is the `*` operator is identical to `**` for `matchPattern: *`, which is logically inconsistent.
And it prevents you from matching against a name without `.`, e.g. `example` but not `example.com`.

**Comment by joestringer** [comment]
Hmm. I agree that it's inconsistent. This is the first I've seen anyone bring up the concern about matching a name without a `.`. However, even if we introduce a brand new field, we cannot change the behavior of `matchPattern: *` as that could break existing network policies. So the end result whether we use a new field or not is that it would not be possible to satisfy this specific use case.

**Comment by erikschul** [comment]
It would be possible to keep the current functionality in `matchPattern`, and new functionality in `matchPattern2`.
If you're already using regex for this, you could also provide an escape hatch as `regexPattern`.

**Comment by joestringer** [comment]
I think I'm still missing what functionality we would lose if the `**` syntax is added to the existing field. There is currently no ability to match on only a single-segment domain `example` right? Because `*` by itself acts as an arbitrary wildcard including the period character.

**Comment by erikschul** [comment]
Correct. Two options:
1) adding `**` to `matchPattern`

+ simple
- `*` is inconsistent in that it behaves like `**` for `matchPattern: *`.
- it does not allow matching single-level domains like `example`

2) using `matchPattern2`
+ doesn't affect existing code
+ clearly communicates a different behavior of the pattern matching
+ `matchPattern2: *` will match only single-level domains like `example`
- may require duplicate code
- slightly ugly

3) same as 1) , but also adding `regexPattern`
+ allows any custom pattern, e.g. single-level domains `[a-zA-Z0-9]+`.
+ regexPattern can be disallowed by strict cluster admins (e.g. using Kyverno) for concerns about regex engine vulnerabilities

Maybe option 3 is the best?

**Comment by phj-incom** [comment]
+1 for option 3.

**Comment by joestringer** [comment]
I would encourage option (1). There's a natural expansion of the existing functionality and the "confusing" part is already part of the API.

Option (2) seems like it's introducing a new field without express need for the new field, as I'm not seeing strong signal from users that supporting single-level domains is critical at this time. Exposing some version of regex as per (3) seems error-prone, we deliberately avoided exposing a regex-like interface in the API to keep the API surface simple, including to reduce the concerns around regex misuse.

**Comment by phj-incom** [comment]
Any progress here? It’s preventing us from using cilium network policies over a traditional firewall in some scenarios.

**Comment by antonengelhardt** [comment]
Hey, it seems like [`MatchPattern` is now available](https://arc.net/l/quote/zpzrckif). But i still really like the idea of `**` to include all subdomains in front of the pattern instead of doing something like this: `*.*.domain.com`. Thanks

**Comment by phj-incom** [comment]
> Hey, it seems like [`MatchPattern` is now available](https://arc.net/l/quote/zpzrckif). But i still really like the idea of `**` to include all subdomains in front of the pattern instead of doing something like this: `*.*.domain.com`. Thanks

Hmm that has been available for a while, and not related to the issue at hand.

**Comment by antonengelhardt** [comment]
Okay, could you add it to the docs then? ;)

**Comment by antonengelhardt** [comment]
Why is it not related? Im just saying that i am in favor of `**` to allow multiple subdomains.

**Comment by joestringer** [comment]
I would encourage you to use emoji reactions on the issue description, or the "Subscribe" button under "Notifications" on the right hand side to endorse/follow this issue. That way we can keep the commentary related to how we design and build this functionality. Thanks for your interest.

**Comment by cthtrifork** [comment]
We are unable to create a pattern for FQDNs such as `prod-registry-k8s-io-eu-central-1.s3.dualstack.eu-central-1.amazonaws.com`

**Comment by AnthonyPoschen** [comment]
just to add to the discussion here, i have recently had to add support for s3 buckets and had some users who liked using dot's in the bucket name, so to add support for them my whitelisting ended up looking like this

```yaml
        - matchPattern: "*.s3.ap-southeast-2.amazonaws.com"
        - matchPattern: "*.*.s3.ap-southeast-2.amazonaws.com"
        - matchPattern: "*.*.*.s3.ap-southeast-2.amazonaws.com"
        - matchPattern: "*.*.*.*.s3.ap-southeast-2.amazonaws.com"
        - matchPattern: "*.*.*.*.*.s3.ap-southeast-2.amazonaws.com"
        - matchPattern: "*.*.*.*.*.*.s3.ap-southeast-2.amazonaws.com"
        - matchPattern: "*.*.*.*.*.*.*.s3.ap-southeast-2.amazonaws.com"
        - matchPattern: "*.*.*.*.*.*.*.*.s3.ap-southeast-2.amazonaws.com"
        - matchPattern: "*.*.*.*.*.*.*.*.*.s3.ap-southeast-2.amazonaws.com"
        - matchPattern: "*.*.*.*.*.*.*.*.*.*.s3.ap-southeast-2.amazonaws.com"
```

and if someone ever made a bucket with more dot's than i added support for id have to extend the list and i would love to just  reach for the following to rid myself of the problem

```yaml
        - matchPattern: "**.s3.ap-southeast-2.amazonaws.com"
```

i have to do this a lot in many places for the AWS api, it feels more error prone in my use cases to not have "**" as an option

**Comment by yangtian9999** [comment]
Does it have any plan to tune to support  this wildcard domain?

**Comment by adamnmcc** [comment]
yes. this makes using it for securing a cluster of CICD runners quite hard when i need to go several levels deep for APIs and AWS resources. 

if this could be re-visited it would be great.

**Comment by pjablonski123** [comment]
Can we conclude a security discussion?

There is a quite high demand of ** hence a PR is recreated here https://github.com/cilium/cilium/pull/42871

**Comment by joestringer** [comment]
Is there a link to the security discussion somewhere? Is it just the subdomain hijacking concern? If yes, it would help to have a writeup that describes how that attack might work and any mitigations we might have or guidance for users to mitigate that risk.

**Comment by pjablonski123** [comment]
> Is there a link to the security discussion somewhere? Is it just the subdomain hijacking concern? If yes, it would help to have a writeup that describes how that attack might work and any mitigations we might have or guidance for users to mitigate that risk.

@joestringer There was a discussion here: https://github.com/cilium/cilium/pull/28690#issuecomment-1777674539

and an agreement to add a security note about a risk: https://github.com/cilium/cilium/pull/28690#issuecomment-1780152093

**Comment by joestringer** [comment]
I see. I don't see the security note commentary in the tree so that seems to be the next step? Seems like we could submit the text in [this comment](https://github.com/cilium/cilium/pull/28690#issuecomment-1779409658) minus the `**` syntax references, get that in & ready for backport. Then afterwards we can rebase the PR on top and extend the wording to add the references to the `**` syntax. Does that make sense?

**Comment by pjablonski123** [comment]
The security note was here (as you already referred to by the link): https://github.com/cilium/cilium/pull/28690#issuecomment-1779409658
and it is included in the files https://github.com/cilium/cilium/pull/42871/files (Documentation/security/policy/language.rst lines 654-662)

**Comment by pjablonski123** [comment]
> minus the ** syntax references, get that in & ready for backport. 

So, just to remove ** syntax references from the note, other code related references leave intact, like this?

> Wildcards as described above should be used with caution, as they could allow outward traffic from a Cilium cluster to unexpected destinations; wildcards in particular could have far-reaching effects. Before implementing policies which use such wildcards, users should evaluate their own threat model and the likelihood of a target domain or sub-domain being configured or hijacked to exfiltrate information from the Cilium deployment. In situations where users do not use security measures like authentication codes to protect their DNS entries or do not have full control of the target domain or sub-domain, usage of such wildcards is generally discouraged.

**Comment by R-Studio** [comment]
Any updates on this?

**Comment by repairman29** [comment]
🎯 **PING RECEIVED!**

**@repairman29** has claimed this bounty via [Echeo](https://echeo.io).

**Match Score:** 85%  
**Status:** Ready to start working!

[View conversation in Echeo](https://echeo.io/messages) to coordinate.

**Comment by repairman29** [comment]
Hey there! This looks like something I could help with. I'd be happy to take a look and see what we can do.

*(Found via [Echeo](https://echeo.io))*

**Comment by therealsamlin** [comment]
I believe this issue is addressed in PR https://github.com/cilium/cilium/pull/43420

**Comment by joestringer** [comment]
Yep! This is now part of v1.19.0-rc.0, to be part of the final v1.19.0 release coming out early February. Please try it out :) And thanks all who helped with the feature.

**Comment by pjablonski123** [comment]
The main goal of this enhancement will be achieved. Thank you all for bringing it to the final stage!
